pinot-query-planner/src/main/java/org/apache/pinot/query/type/TypeSystem.java

```diff
@@ -70,4 +70,16 @@ public RelDataType deriveAvgAggType(RelDataTypeFactory typeFactory,
       }
     }
   }
+
+  // overriding the function for handling comparison between different data types.
+  // based on the precision value calcite adds the CAST function to make operands of same type for comparison.
+  @Override public int getDefaultPrecision(SqlTypeName typeName) {
+    switch (typeName) {
+      case DOUBLE:
+        return 16;
+      default:
+        // Following BasicSqlType precision as the default
+        return super.getDefaultPrecision(typeName);
+    }
```

Comment:
i think this is the problem : in JavaTypeFactoryImpl
```
  @Override public Type getJavaClass(RelDataType type) {
...
    if (type instanceof BasicSqlType || type instanceof IntervalSqlType) {
      switch (type.getSqlTypeName()) {
...
      case DOUBLE:
      case FLOAT: // sic
        return type.isNullable() ? Double.class : double.class;
      case REAL:
        return type.isNullable() ? Float.class : float.class;
```

This means `SQL.REAL` is actually java Float. and `SQL.FLOAT` or `SQL.DOUBLE` is actually java Double...

This means all of our interpretation on float is wrong (b/c we equivalent `SQL.FLOAT` as java Float

This is a much bigger problem IMO

Reply:
Well, I changed the `TypeFactory` extending`SqlTypeFactoryImpl` instead of  `JavaTypeFactoryImpl`, and there is no effect on our test suite. Here are the [changes](https://github.com/apache/pinot/pull/10883).  
I will say either our test suite isn't complete to cover complete testing or it doesn't matter if we extend either type. Let me look `TypeCoercion` logic.

Reply:
`SqlTypeFactoryImpl` uses `RelDataTypeSystemImpl` to compute the scale/precision, which behaves the same as JavaTypeFactoryImpl. 

```
...
    case REAL:
      return 7;
    case FLOAT:
    case DOUBLE:
      return 15;
```